man/crashmaint.1
.TH CRASHMAINT 1 1999-08-01 "Johan Billing" "CrashMail"
.SH NAME
crashmaint \- Do maintanence on CrashMail message bases
.SH SYNOPSIS
.B crashmaint
[MAINT] [PACK] [VERBOSE] [SETTINGS filename] [PATTERN pattern]
.SH DESCRIPTION
Deletes old messages according to
.B KEEPNUM
and
.B KEEPDAYS
in
.IR crashmail.prefs .
The program can do two different operations on a messagebase,
.I MAINT
and
.IR PACK .
The meaning of these two modes are different for different messagebase
formats.
.TP
.B *.msg
.I MAINT
deletes messages and
.I PACK
renumbers the area.
.TP
.B JAM
.I MAINT
sets the Deleted flag for the messages.
.I PACK
removes all messages with the Deleted flag from the messagebase.
.PP
Both
.I MAINT
and
.I PACK
can be specified at the same time.
.SH OPTIONS
.TP
.I SETTINGS filename
Specifies that another config file than the default should be used. You can use
the environment variable CMCONFIGFILE to set the default configuration file.
.TP
.I PATTERN pattern
Lets you perform the operations on only some of your areas.
.TP
.I VERBOSE
Gives you a lot of information which you don't really need.
.SH EXAMPLES
crashmaint MAINT PACK PATTERN R20_AMIGA*
.SH "SEE ALSO"
.BR crashmail (1)
.\"BR crashmail.prefs (5)
.\"SH FILES
.\"SH BUGS
.SH AUTHOR
CrashMail is written by Johan Billing <billing@df.lth.se>
.PP
This manual page was written by Peter Karlsson <peter@softwolves.pp.se>


doc/History.txt
Changes in version 0.4

 - First public release
 
Changes in version 0.42

 - Fixed bug that occurred when sending mail to nodes with long addresses 
   (e.g. 9999:8888/7777.6666)
	
 - Fixed an error message in the outbound handling
 
 - Fixed a bug that occurred when auto-adding pass-through areas

 - Fixed a small memory leak

 - CrashMail got caught in an infinite loop when an unknown nodelist format
   was specified in the config. This is now fixed.
	
 - CrashMail no longer segfaults at startup when pass-through areas are used!
 
 - CrashMail now adds packets to bundles in the correct order
 
 - CrashMail would under some circumstances use the same bundle for more than
   one node. This has been fixed and a mechanism that checks that the same
   bundle is not used for more than one node has been added.

 - Routing did not work properly when BOUNCEPOINTS was turned on. Fixed. 

 - Removed an unnecessary linebreak from the bounce messages.

 - New configuration keyword: NOMAXOUTBOUNDZONE. See example.prefs for more
   information.

 - CrashList did not handle Pvt nodes correctly. Fixed.

 - There was a small mistake in the example configuration file. The keyword is
   REMOTEAF, not REMOTEAREAFIX. 

Changes in version 0.5

 CrashMail:

 - Because of a bug in the pattern matching, an empty string would match any
   pattern. Fixed.
	
 - Areas belonging to the same group were not listed together in the area lists
   generated by the Areafix.
	 	
 - CrashMail showed the same error message twice when the packer returned an
   error when unarchiving a bundle. Fixed.
	 
 - If the highwater mark was lower than the lowest messagenumber in a JAM area,
   CrashMail would still try to export messages beginning from the highwater
   mark. Unfortunately this would cause a crash because of another bug
   that appeared when CrashMail tried to export an unreadable message.
   All this has now been fixed.
	
 - CrashMail no longer stops exporting when it encounters an unreadable message

 - CrashMail no longer adds its own node number to the path if it already 
   exists as the last node in the path lines.

 - CrashMail could sometimes create circular reply links and get caught in 
   infinite loops when dupes were imported to a JAM messagebase. Fixed.

 - New config option: ALLOWKILLSENT

 - New command-line keyword for CrashMail: TOSSDIR
	
 - New command-line keyword for CrashMail: NOSECURITY. If this keyword is
   used, all packets will be tossed without security checks. This is mainly 
   intended to be used together with TOSSFILE/TOSSDIR and packets
   generated by CrashWrite.
	
 CrashMaint:

 - CrashMaint no longer checks the ActiveMsgs counter in the JAM messagebase
   to determine if the area needs to be packed. This check did not always work,
   among other things because all programs did not update the counter
   properly. JAM areas are now always packed when "crashmaint pack" is
   used.

 - CrashMaint now rebuilds the ActiveMsgs counter when packing a JAM area.

 - CrashMaint did not update the messagebase header when removing all messages
   from a JAM area. This would cause some messages not to be exported when 
   highwater marks were used. (If you suspect that this has happened to
   you, remove all *.cmhw files.)
	
 - Improved the error handling a bit in CrashMaint. Now you get an error if 
   an area couldn't be updated since it was in use by another program.

 CrashWrite:
 
 - Included CrashWrite, a new tool for CrashMail. CrashWrite reads a text
   file and then creates a packet file that can be tossed by CrashMail.

 CrashExport:
 
 - CrashAreasBBS and CrashForward have been replaced by the new program
   CrashExport. In addition to the old formats, CrashExport can also
   create arealists in GoldED format. The old version of CrashAreasBBS had 
   some trouble with pass-through areas, this is gone in CrashExport.

Changes in version 0.52

 CrashMail:

 - Three new flags for the NODE keyword: CRASH, DIRECT and HOLD. With these
   flags, you can set the priority for packed and unpacked netmail.

 - When the packet directory was missing, the error message said that the
   outbound directory was missing. This could be very confusing. Fixed.

 - Fixed a Y2K-related bug in the JAM import function. 

 - Node patterns with REGION or HUB did not work for points. Fixed.

 - Added the new %-code %R for the ROBOTNAME keyword. %R gives the message
   in RFC-format with the addresses in RFC-format (name@pX.fX.nX.zX.domain).

 - When importing messages, CrashMail now checks for robot and AreaFix names
   before messages are remapped. If you want to remap messages to AreaFix, you 
   have to make sure that no AreaFix names are configured. REMAPNODE still
   works as before. 
	

tools/crashmaint.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <signal.h>
#include <errno.h>

#include <jamlib/jam.h>

#define NO_TYPEDEF_UCHAR
#define NO_TYPEDEF_ULONG
#define NO_TYPEDEF_USHORT

#include <shared/parseargs.h>
#include <shared/jblist.h>
#include <shared/jbstrcpy.h>
#include <shared/path.h>

#include <oslib/os.h>
#include <oslib/osmem.h>
#include <oslib/osfile.h>
#include <oslib/osdir.h>
#include <oslib/ospattern.h>
#include <oslib/osmisc.h>

#include <shared/storedmsg.h>
#include <shared/fidonet.h>

#define VERSION "1.2"

#ifdef PLATFORM_AMIGA
uchar *ver="$VER: CrashMaint "VERSION" ("__COMMODORE_DATE__")";
#endif

struct Area
{
   struct Area *Next;
   uchar Tagname[80];
   uchar Path[80];
	uchar Messagebase[20];
   ulong KeepNum,KeepDays;
};

struct jbList AreaList;

struct Messagebase
{
	uchar *Name;
	bool (*processfunc)(struct Area *area,bool maint,bool pack,bool verbose);
};

#ifdef MSGBASE_MSG
bool ProcessAreaMSG(struct Area *area,bool maint, bool pack, bool verbose);
#endif

#ifdef MSGBASE_JAM
bool ProcessAreaJAM(struct Area *area,bool maint, bool pack, bool verbose);
#endif

struct Messagebase Messagebases[] = 
{ 
#ifdef MSGBASE_JAM
	{ "JAM", ProcessAreaJAM },
#endif

#ifdef MSGBASE_MSG
	{ "MSG", ProcessAreaMSG },
#endif								  

	{ NULL,  NULL       } 
};

#define ARG_MAINT      0
#define ARG_PACK       1
#define ARG_VERBOSE    2
#define ARG_SETTINGS   3
#define ARG_PATTERN    4

struct argument args[] =
   { { ARGTYPE_BOOL,   "MAINT",        0, NULL },
     { ARGTYPE_BOOL,   "PACK",         0, NULL },
     { ARGTYPE_BOOL,   "VERBOSE",      0, NULL },
     { ARGTYPE_STRING, "SETTINGS",     0, NULL },
     { ARGTYPE_STRING, "PATTERN",      0, NULL },
     { ARGTYPE_END,     NULL,          0, 0    } };

bool ctrlc;

void breakfunc(int x)
{
   ctrlc=TRUE;
}

/******************** *.msg *********************/

#ifdef MSGBASE_MSG

struct Msg
{
   struct Msg *Next;
   ulong Num,NewNum,Day;
};

struct jbList MsgList;

int Compare(const void *a1,const void *a2)
{
   struct Msg **m1,**m2;

   m1=(struct Msg **)a1;
   m2=(struct Msg **)a2;

   if((*m1)->Num > (*m2)->Num) return(1);
   if((*m1)->Num < (*m2)->Num) return(-1);
   return(0);
}

bool Sort(struct jbList *list)
{
   struct Msg *msg,**buf,**work;
   ulong nc;

   nc=0;

   for(msg=(struct Msg *)list->First;msg;msg=msg->Next)
      nc++;

   if(nc == 0)
      return(TRUE);

   if(!(buf=(struct Msg **)osAlloc(nc*sizeof(struct StatsNode *))))
      return(FALSE);

   work=buf;

   for(msg=(struct Msg *)list->First;msg;msg=msg->Next)
      *work++=msg;

   qsort(buf,nc,4,Compare);

   jbNewList(list);

   for(work=buf;nc--;)
      jbAddNode(list,(struct jbNode *)*work++);

   osFree(buf);

   return(TRUE);
}

void MakeFidoDate(time_t tim,uchar *dest)
{
   struct tm *tp;
   time_t t;
   uchar *monthnames[]={"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec","???"};

   t=tim;
   tp=localtime(&t);

   sprintf(dest,"%02d %s %02d  %02d:%02d:%02d",
      tp->tm_mday,
      monthnames[tp->tm_mon],
      tp->tm_year % 100,
      tp->tm_hour,
      tp->tm_min,
      tp->tm_sec);
}

uchar *scanfuncarea;
bool nomem;

void scanfunc(uchar *str)
{
   uchar buf[200];
   struct osFileEntry *fe;
   ulong num,day;
   struct Msg *msg;

   if(strlen(str) < 5)
      return;

   if(stricmp(&str[strlen(str)-4],".msg")!=0)
      return;
                                                                             
   if(atol(str) < 2)
      return;

   MakeFullPath(scanfuncarea,str,buf,200);

   if(!(fe=osGetFileEntry(buf)))
      return;

   num=atol(str);   
   day=fe->Date / (24*60*60);

   osFree(fe);

   if(!(msg=(struct Msg *)osAlloc(sizeof(struct Msg))))
   {
      nomem=TRUE;
      return;
   }

   jbAddNode(&MsgList,(struct jbNode *)msg);

   msg->Num=num;
   msg->Day=day;
}

bool ProcessAreaMSG(struct Area *area,bool maint, bool pack, bool verbose)
{
   ulong today,num,del,highwater,oldhighwater;
   struct Msg *msg;
   uchar buf[200],newbuf[200],buf2[100];
   struct StoredMsg StoredMsg;
   osFile fh;

   highwater=0;
   oldhighwater=0;

   jbNewList(&MsgList);

   printf("Processing %s...\n",area->Tagname);

   scanfuncarea=area->Path;

   if(!(osScanDir(area->Path,scanfunc)))
   {
		ulong err=osError();
      printf(" Error: Couldn't scan directory %s\n",area->Path);
		printf(" Error: %s\n",osErrorMsg(err));
      jbFreeList(&MsgList);
      return(TRUE);
   }

   if(nomem)
   {
      printf("Out of memory\n");
      jbFreeList(&MsgList);
      return(FALSE);
   }

   if(!Sort(&MsgList))
   {
      printf("Out of memory\n");
      jbFreeList(&MsgList);
      return(FALSE);
   }

   if(!MsgList.First)
   {
      printf(" Area is empty\n");
      return(TRUE);
   }

   if(ctrlc)
   {
      jbFreeList(&MsgList);
      return(TRUE);
   }

   MakeFullPath(area->Path,"1.msg",buf,200);

   if((fh=osOpen(buf,MODE_OLDFILE)))
   {
      if(osRead(fh,&StoredMsg,sizeof(struct StoredMsg))==sizeof(struct StoredMsg))
      {
         highwater=StoredMsg.ReplyTo;
         oldhighwater=StoredMsg.ReplyTo;
      }
      osClose(fh);
   }

   if(maint && area->KeepNum!=0)
   {
      num=0;

      for(msg=(struct Msg *)MsgList.First;msg;msg=msg->Next)
         num++;

      msg=(struct Msg *)MsgList.First;
      del=0;

      while(num>area->KeepNum && !ctrlc)
      {
         while(msg->Num==0)
            msg=msg->Next;

         sprintf(buf2,"%lu.msg",msg->Num);
         MakeFullPath(area->Path,buf2,buf,200);

         if(msg->Num == highwater)
            highwater=0;

         if(verbose)
            printf(" Deleting message #%lu by number\n",msg->Num);

         osDelete(buf);

         msg->Num=0;
         num--;
         del++;
      }

      if(ctrlc)
      {
         jbFreeList(&MsgList);
         return(TRUE);
      }

      printf(" %lu messages deleted by number, %lu messages left\n",del,num);
   }

   if(maint && area->KeepDays!=0)
   {
      del=0;
      num=0;
		
		today=time(NULL) / (24*60*60);

      for(msg=(struct Msg *)MsgList.First;msg && !ctrlc;msg=msg->Next)
      {
         if(today - msg->Day > area->KeepDays && msg->Num!=0)
         {
            sprintf(buf2,"%lu.msg",msg->Num);
            MakeFullPath(area->Path,buf2,buf,200);

            if(msg->Num == highwater)
               highwater=0;

            if(verbose)
               printf(" Deleting message #%lu by date\n",msg->Num);

            osDelete(buf);
            
      	    msg->Num=0;
            del++;
         }
         else
         {
            num++;
         }
      }

      if(ctrlc)
      {
         jbFreeList(&MsgList);
         return(TRUE);
      }

      printf(" %lu messages deleted by date, %lu messages left\n",del,num);
   }

   if(pack)
   {
      num=2;

      msg=(struct Msg *)MsgList.First;

      while(msg && !ctrlc)
      {
         while(msg && msg->Num==0)
            msg=msg->Next;

         if(msg)
         {
            msg->NewNum=num++;
            msg=msg->Next;
         }
      }
 
      for(msg=(struct Msg *)MsgList.First;msg && !ctrlc;msg=msg->Next)
         if(msg->Num!=0 && msg->Num!=msg->NewNum)
         {
            sprintf(buf2,"%lu.msg",msg->Num);
				MakeFullPath(area->Path,buf2,buf,200);

            sprintf(buf2,"%lu.msg",msg->NewNum);
				MakeFullPath(area->Path,buf2,newbuf,200);

            if(highwater == msg->Num)
               highwater=msg->NewNum;

            if(verbose)
               printf(" Renaming message %lu to %lu\n",msg->Num,msg->NewNum);

            osRename(buf,newbuf);
         }

      if(ctrlc)
      {
         jbFreeList(&MsgList);
         return(TRUE);
      }

      printf(" Area renumbered\n");
   }

   jbFreeList(&MsgList);

   if(highwater!=oldhighwater)
   {
      strcpy(StoredMsg.From,"CrashMail II");
      strcpy(StoredMsg.To,"All");
      strcpy(StoredMsg.Subject,"HighWater mark");
      MakeFidoDate(time(NULL),StoredMsg.DateTime);

      StoredMsg.TimesRead=0;
      StoredMsg.DestNode=0;
      StoredMsg.OrigNode=0;
      StoredMsg.Cost=0;
      StoredMsg.OrigNet=0;
      StoredMsg.DestNet=0;
      StoredMsg.DestZone=0;
      StoredMsg.OrigZone=0;
      StoredMsg.OrigPoint=0;
      StoredMsg.DestPoint=0;
      StoredMsg.ReplyTo=highwater;
      StoredMsg.Attr=FLAG_SENT | FLAG_PVT;
      StoredMsg.NextReply=0;

      MakeFullPath(area->Path,"1.msg",buf,200);

      if((fh=osOpen(buf,MODE_NEWFILE)))
      {
         osWrite(fh,&StoredMsg,sizeof(struct StoredMsg));
         osWrite(fh,"",1);
         osClose(fh);
      }
   }

   return(TRUE);
}

#endif

/*************************** JAM ************************/

#ifdef MSGBASE_JAM

long jam_utcoffset = 0xbaadf00d;

bool ProcessAreaJAM(struct Area *area,bool maint, bool pack, bool verbose)
{
   ulong today,active,basenum,total,del,num,day;
   s_JamBase *Base_PS,*NewBase_PS;
   s_JamBaseHeader BaseHeader_S;
   s_JamMsgHeader	Header_S;
   s_JamSubPacket*	SubPacket_PS;
	int res,res1,res2;
	uchar buf[200],oldname[200],tmpname[200];
	bool firstwritten;
	uchar *msgtext;

   /* Some timezone tricks */

   if(jam_utcoffset == 0xbaadf00d) 
   {
      time_t t1,t2;
      struct tm *tp;

      t1=time(NULL);
      tp=gmtime(&t1);
      tp->tm_isdst=-1;
      t2=mktime(tp);
      jam_utcoffset=t2-t1;
   }
			
   printf("Processing %s...\n",area->Tagname);

   if(JAM_OpenMB(area->Path,&Base_PS))
   {
      printf(" Failed to open messagebase \"%s\"\n",area->Path);
		return(TRUE);
	}
	
	if(JAM_LockMB(Base_PS,10))
	{
		printf(" Timeout when trying to lock messagebase \"%s\"\n",area->Path);
		JAM_CloseMB(Base_PS);
		return(TRUE);
	}

   if(JAM_ReadMBHeader(Base_PS,&BaseHeader_S))
   {
      printf(" Failed to read header of messagebase \"%s\"\n",area->Path);
		JAM_UnlockMB(Base_PS);
		JAM_CloseMB(Base_PS);
      return(TRUE);
   }

   if(JAM_GetMBSize(Base_PS,&total))
   {
      printf(" Failed to get size of messagebase \"%s\"\n",area->Path);
		JAM_UnlockMB(Base_PS);
      JAM_CloseMB(Base_PS);
      return(TRUE);
   }

   basenum=BaseHeader_S.BaseMsgNum;
   active=BaseHeader_S.ActiveMsgs;
	
   if(total == 0)
   {
      printf(" Area is empty\n");
		JAM_UnlockMB(Base_PS);
		JAM_CloseMB(Base_PS);
      return(TRUE);
   }

   if(maint && area->KeepNum!=0)
   {		
		num=0;
		del=0;
		
		while(num < total && active > area->KeepNum && !ctrlc)
		{
		   res=JAM_ReadMsgHeader(Base_PS,num,&Header_S,NULL);

         if(res == 0)
			{
				/* Read success */
				
			   if(!(Header_S.Attribute & MSG_DELETED))
				{
					/* Not already deleted */

		         if(verbose)
      		      printf(" Deleting message #%lu by number\n",basenum+num);
	
					Header_S.Attribute |= MSG_DELETED;
					JAM_ChangeMsgHeader(Base_PS,num,&Header_S);

					BaseHeader_S.ActiveMsgs--;
					JAM_WriteMBHeader(Base_PS,&BaseHeader_S);
					
					active--;
					del++;
				}
			}
			
			num++;
		}

      if(ctrlc)
      {
			JAM_UnlockMB(Base_PS);
			JAM_CloseMB(Base_PS);
         return(TRUE);
      }

      printf(" %lu messages deleted by number, %lu messages left\n",del,active);
   }

   if(maint && area->KeepDays!=0)
   {
      del=0;
      num=0;
		
      today=(time(NULL)-jam_utcoffset) / (24*60*60);

		while(num < total && !ctrlc)
		{
		   res=JAM_ReadMsgHeader(Base_PS,num,&Header_S,NULL);

         if(res == 0)
			{
				/* Read success */

				day=Header_S.DateReceived / (24*60*60);
				
				if(day == 0)
					day=Header_S.DateProcessed / (24*60*60);
				
				if(day == 0)
					day=Header_S.DateWritten / (24*60*60);

			   if(today-day > area->KeepDays && !(Header_S.Attribute & MSG_DELETED))
				{
					/* Not already deleted and too old*/

		         if(verbose)
      		      printf(" Deleting message #%lu by date\n",basenum+num);
	
					Header_S.Attribute |= MSG_DELETED;
					JAM_ChangeMsgHeader(Base_PS,num,&Header_S);

					BaseHeader_S.ActiveMsgs--;
					JAM_WriteMBHeader(Base_PS,&BaseHeader_S);
					
					del++;
					active--;
				}
			}
			
			num++;
		}

      if(ctrlc)
      {
			JAM_UnlockMB(Base_PS);
			JAM_CloseMB(Base_PS);
         return(TRUE);
      }

      printf(" %lu messages deleted by date, %lu messages left\n",del,active);
   }

   if(pack)
   {
		strcpy(buf,area->Path);
		strcat(buf,".cmtemp");

	   if(JAM_CreateMB(buf,1,&NewBase_PS))
   	{
      	printf(" Failed to create new messagebase \"%s\"\n",buf);
			JAM_UnlockMB(Base_PS);
			JAM_CloseMB(Base_PS);
			return(TRUE);
		}

		if(JAM_LockMB(NewBase_PS,10))
		{
			printf(" Timeout when trying to lock messagebase \"%s\"\n",buf);
			JAM_UnlockMB(Base_PS);
			JAM_CloseMB(Base_PS);
			JAM_CloseMB(NewBase_PS);
			JAM_RemoveMB(NewBase_PS,buf);
			return(TRUE);
		}

		/* Copy messages */

      del=0;
      num=0;
		firstwritten=FALSE;

		BaseHeader_S.ActiveMsgs=0;
				
		while(num < total && !ctrlc)
		{
		   res=JAM_ReadMsgHeader(Base_PS,num,&Header_S,NULL);

         if(res)
			{
				if(res == JAM_NO_MESSAGE)
				{
					if(firstwritten) 
					{
						JAM_AddEmptyMessage(NewBase_PS);
					}
					else 
					{
						BaseHeader_S.BaseMsgNum++;
						del++;
					}
				}
				else
				{
		      	printf(" Failed to read message %ld, cannot pack messagebase\n",num+basenum);
					JAM_UnlockMB(Base_PS);
					JAM_CloseMB(Base_PS);
					JAM_UnlockMB(NewBase_PS);
					JAM_CloseMB(NewBase_PS);
					JAM_RemoveMB(NewBase_PS,buf);
					return(TRUE);
				}
			}
			else
			{
				if(Header_S.Attribute & MSG_DELETED)
				{
					if(firstwritten) 
					{
						JAM_AddEmptyMessage(NewBase_PS);
					}
					else
					{
						BaseHeader_S.BaseMsgNum++;					
						del++;
					}
				}
				else
				{
               if(!firstwritten)
               {
                  /* Set basenum */

                  res=JAM_WriteMBHeader(NewBase_PS,&BaseHeader_S);

                  if(res)
                  {
                     printf(" Failed to write messagebase header, cannot pack messagebase\n");
                     JAM_UnlockMB(Base_PS);
                     JAM_CloseMB(Base_PS);
                     JAM_UnlockMB(NewBase_PS);
                     JAM_CloseMB(NewBase_PS);
                     JAM_RemoveMB(NewBase_PS,buf);
                     return(TRUE);
                  }

                  firstwritten=TRUE;
               }

					/* Read header with all subpackets*/

					res=JAM_ReadMsgHeader(Base_PS,num,&Header_S,&SubPacket_PS);

				   if(res)
				   {
			      	printf(" Failed to read message %ld, cannot pack messagebase\n",num+basenum);
						JAM_UnlockMB(Base_PS);
						JAM_CloseMB(Base_PS);
						JAM_UnlockMB(NewBase_PS);
						JAM_CloseMB(NewBase_PS);
						JAM_RemoveMB(NewBase_PS,buf);
						return(TRUE);
					}

				   /* Read message text */

				   msgtext=NULL;

				   if(Header_S.TxtLen)
				   {
				      if(!(msgtext=osAlloc(Header_S.TxtLen)))
				      {
							printf("Out of memory\n");
				         JAM_DelSubPacket(SubPacket_PS);
							JAM_UnlockMB(Base_PS);
							JAM_CloseMB(Base_PS);
							JAM_UnlockMB(NewBase_PS);
							JAM_CloseMB(NewBase_PS);
							JAM_RemoveMB(NewBase_PS,buf);
				         return(FALSE);
						}			

				      res=JAM_ReadMsgText(Base_PS,Header_S.TxtOffset,Header_S.TxtLen,msgtext);

				      if(res)
				      {
				      	printf(" Failed to read message %ld, cannot pack messagebase\n",num+basenum);
				         JAM_DelSubPacket(SubPacket_PS);
							JAM_UnlockMB(Base_PS);
							JAM_CloseMB(Base_PS);
							JAM_UnlockMB(NewBase_PS);
							JAM_CloseMB(NewBase_PS);
							JAM_RemoveMB(NewBase_PS,buf);
					      return(TRUE);
						}
			      }

					/* Write new message */

				   res=JAM_AddMessage(NewBase_PS,&Header_S,SubPacket_PS,msgtext,Header_S.TxtLen);

					if(msgtext) osFree(msgtext);
				   JAM_DelSubPacket(SubPacket_PS);

					BaseHeader_S.ActiveMsgs++;

					if(res)
					{
				      printf(" Failed to copy message %ld (disk full?), cannot pack messagebase\n",num+basenum);
						JAM_UnlockMB(Base_PS);
						JAM_CloseMB(Base_PS);
						JAM_UnlockMB(NewBase_PS);
						JAM_CloseMB(NewBase_PS);
						JAM_RemoveMB(NewBase_PS,buf);
					   return(TRUE);
					}
				}
			}
			
			num++;
		}

		/* Write back header */

		BaseHeader_S.ModCounter++;

      res=JAM_WriteMBHeader(NewBase_PS,&BaseHeader_S);

     	if(res)
     	{
        	printf(" Failed to write messagebase header, cannot pack messagebase\n");
        	JAM_UnlockMB(Base_PS);
        	JAM_CloseMB(Base_PS);
        	JAM_UnlockMB(NewBase_PS);
			JAM_CloseMB(NewBase_PS);
        	JAM_RemoveMB(NewBase_PS,buf);
        	return(TRUE);
      }

		JAM_UnlockMB(Base_PS);
		JAM_CloseMB(Base_PS);

		JAM_UnlockMB(NewBase_PS);
		JAM_CloseMB(NewBase_PS);

		if(ctrlc)
		{
			JAM_RemoveMB(NewBase_PS,buf);
         return(TRUE);
      }
		
		/* This could not be done with JAMLIB... */

	   sprintf(oldname,"%s%s",area->Path,EXT_HDRFILE);
  	   sprintf(tmpname,"%s.cmtemp%s",area->Path,EXT_HDRFILE);
		res1=osDelete(oldname);
		res2=osRename(tmpname,oldname);

		if(res1 && res2)
		{
		   sprintf(oldname,"%s%s",area->Path,EXT_TXTFILE);
  		   sprintf(tmpname,"%s.cmtemp%s",area->Path,EXT_TXTFILE);
			res1=osDelete(oldname);
			res2=osRename(tmpname,oldname);
		}
		
		if(res1 && res2)
		{
	   	sprintf(oldname,"%s%s",area->Path,EXT_IDXFILE);
  	   	sprintf(tmpname,"%s.cmtemp%s",area->Path,EXT_IDXFILE);
			res1=osDelete(oldname);
			res2=osRename(tmpname,oldname);
		}
		
		if(res1 && res2)
		{
	   	sprintf(oldname,"%s%s",area->Path,EXT_LRDFILE);
  	   	sprintf(tmpname,"%s.cmtemp%s",area->Path,EXT_LRDFILE);
			/* Keep lastread file */
			res2=osDelete(tmpname);
		}
		
		if(!res1 || !res2)
		{
			printf(" Failed to update area. The area might be in use by another program.\n");
			return(FALSE);
		}

      printf(" %ld deleted messages removed from messagebase\n",del);
   }
	else
	{
		JAM_UnlockMB(Base_PS);
		JAM_CloseMB(Base_PS);
	}
	
   return(TRUE);
}

#endif

/************************** end of messagebases *******************/

uchar cfgbuf[4000];

bool ReadConfig(uchar *file)
{
   osFile fh;
   uchar cfgword[20];
   uchar tag[80],aka[80],path[80],mb[20];
   struct Area *tmparea,*LastArea;
   ulong jbcpos;

   if(!(fh=osOpen(file,MODE_OLDFILE)))
	{
		ulong err=osError();
      printf("Failed to open file %s for reading\n",file);
		printf("Error: %s\n",osErrorMsg(err));
      return(FALSE);
	}
	
	LastArea=NULL;

   while(osFGets(fh,cfgbuf,4000))
   {
      jbcpos=0;
      jbstrcpy(cfgword,cfgbuf,20,&jbcpos);

      if(stricmp(cfgword,"KEEPDAYS")==0 && LastArea)
      {
         if(jbstrcpy(tag,cfgbuf,80,&jbcpos))
            LastArea->KeepDays=atol(tag);
      }

      if(stricmp(cfgword,"KEEPNUM")==0 && LastArea)
      {
         if(jbstrcpy(tag,cfgbuf,80,&jbcpos))
            LastArea->KeepNum=atol(tag);
      }

      if(stricmp(cfgword,"AREA")==0 || stricmp(cfgword,"NETMAIL")==0 || stricmp(cfgword,"LOCALAREA")==0)
      {
         jbstrcpy(tag,cfgbuf,80,&jbcpos); 
         jbstrcpy(aka,cfgbuf,80,&jbcpos);

         if(stricmp(tag,"DEFAULT")!=0 && strnicmp(tag,"DEFAULT_",8)!=0)
         {
				if(jbstrcpy(mb,cfgbuf,20,&jbcpos))
				{
				   jbstrcpy(path,cfgbuf,80,&jbcpos);
					
               if(!(tmparea=(struct Area *)osAllocCleared(sizeof(struct Area))))
               {
         		   printf("Out of memory\n");
                  osClose(fh);
                  return(FALSE);
               }                  
					
					jbAddNode(&AreaList,(struct jbNode *)tmparea);
               LastArea=tmparea;

               strcpy(tmparea->Tagname,tag);
               strcpy(tmparea->Messagebase,mb);
               strcpy(tmparea->Path,path);
            }
         }
      }
   }

   osClose(fh);
   return(TRUE);
}

int main(int argc, char **argv)
{
   struct Area *area;
   uchar *cfg;
   bool maint,pack,verbose;
	int i;
		
   signal(SIGINT,breakfunc);

   if(!osInit())
      exit(OS_EXIT_ERROR);

   if(argc > 1 &&
	  (strcmp(argv[1],"?")==0      ||
		strcmp(argv[1],"-h")==0     ||
		strcmp(argv[1],"--help")==0 ||
		strcmp(argv[1],"help")==0 ||
		strcmp(argv[1],"/h")==0     ||
		strcmp(argv[1],"/?")==0 ))
   {
      printargs(args);
      osEnd();
      exit(OS_EXIT_OK);
   }

   if(!parseargs(args,argc,argv))
   {
      osEnd();
      exit(OS_EXIT_ERROR);
   }

   jbNewList(&AreaList);

	maint=FALSE;
	pack=FALSE;
	verbose=FALSE;

   if(args[ARG_MAINT].data)
		maint=TRUE;
	
	if(args[ARG_PACK].data)
		pack=TRUE;

   if(args[ARG_VERBOSE].data)
		verbose=TRUE;

   if(!maint && !pack)
   {
      printf("Nothing to do.\n");
      osEnd();
      exit(OS_EXIT_OK);
   }

   if(args[ARG_PATTERN].data)
   {
      if(!(osCheckPattern((uchar *)args[ARG_PATTERN].data)))
      {
         printf("Invalid pattern \"%s\"\n",(uchar *)args[ARG_PATTERN].data);
         osEnd();
         exit(OS_EXIT_ERROR);
      }
   }
  
	cfg=getenv(OS_CONFIG_VAR);

   if(!cfg)
		cfg=OS_CONFIG_NAME;

   if(args[ARG_SETTINGS].data)
      cfg=(uchar *)args[ARG_SETTINGS].data;

   if(!(ReadConfig(cfg)))
   {
      jbFreeList(&AreaList);
      osEnd();
      exit(OS_EXIT_ERROR);
   }

   for(area=(struct Area *)AreaList.First;area && !ctrlc;area=area->Next)
   {
      bool match;

      match=FALSE;

      if(!args[ARG_PATTERN].data)
         match=TRUE;

      else if(osMatchPattern((uchar *)args[ARG_PATTERN].data,area->Tagname))
         match=TRUE;

      if(match)
      {
			for(i=0;Messagebases[i].Name;i++)
				if(stricmp(Messagebases[i].Name,area->Messagebase)==0) break;

			if(Messagebases[i].processfunc)
         {
            if(!Messagebases[i].processfunc(area,maint,pack,verbose))
               exit(OS_EXIT_ERROR);
         }
			else
			{
				printf("Cannot process area %s, messagebase %s not supported by CrashMaint\n",
					area->Tagname,
					area->Messagebase);
			}
      }
   }

   if(ctrlc)
      printf("*** User Break ***\n");

   jbFreeList(&AreaList);
   osEnd();

   exit(OS_EXIT_OK);
}
